src/analysis/normed_space/units.lean

```diff
@@ -0,0 +1,82 @@
+/-
+Copyright (c) 2020 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth
+-/
+import analysis.normed_space.basic
+import analysis.specific_limits
+import tactic.noncomm_ring
+
+/-!
+# The group of units of a complete normed ring
+
+This file contains the basic theory for the group of units (invertible elements) of a complete
+normed ring (Banach algebras being a notable special case).
+
+## Main results
+
+The constructions `one_sub`, `add` and `unit_of_nearby` state, in varying forms, that perturbations
+of a unit are units.  The latter two are not stated in their optimal form; more precise versions
+would use the spectral radius.
+
+The main result is `is_open`:  the group of units of a complete normed ring is an open subset of the
+ring.
+
+-/
+
+noncomputable theory
+
+namespace units
+variables {α : Type*} [normed_ring α] [complete_space α]
+
+/-- In a complete normed ring, a perturbation of `1` by an element `t` of distance less than `1`
+from `1` is a unit.  Here we construct its `units` structure.  -/
+def one_sub (t : α) (h : ∥t∥ < 1) : units α :=
+{ val := 1 - t,
+  inv := ∑' (n : ℕ), t ^ n,
+  val_inv := mul_neg_geom_series t h,
+  inv_val := geom_series_mul_neg t h }
+
+@[simp] lemma one_sub_coe (t : α) (h : ∥t∥ < 1) : ((one_sub t h) : α) = 1 - t := rfl
+
+/-- In a complete normed ring, a perturbation of a unit `x` by an element `t` of distance less than
+`∥x⁻¹∥⁻¹` from `x` is a unit.  Here we construct its `units` structure. -/
+def add [nonzero α] (x : units α) (t : α) (h : ∥t∥ < ∥((x⁻¹:units α):α)∥⁻¹) : units α :=
```

Comment:
I don't see why you need `nonzero α` here: otherwise the norm in the proof is zero, and so the proof holds as `0 < 1` (also the assumption `h` can never be satisfied, and you can deduce anything from `false`). Then this means this assumption can also be removed from all the lemmas below. Am I missing something?

Reply:
This was just my squeamishness at having `∥0∥⁻¹` in a lemma statement ... I will make the change.